src/app/Library/CrudPanel/Uploads/RegisterUploadEvents.php

```diff
@@ -0,0 +1,185 @@
+        // Changing `subfields` properties will call the macros again. We prevent duplicate entries by checking
+        // if the uploader is already registered.
+        if (app('UploadersRepository')->isUploadRegistered($uploader->getRepeatableContainerName(), $uploader)) {
+            return;
+        }
+
+        $model = $subfield['baseModel'] ?? $crudObject['model'] ?? get_class($this->crudObject->crud()->getModel());
+
+        if (isset($crudObject['relation_type']) && $crudObject['entity'] !== false) {
+            $uploader = $uploader->relationship(true);
+        }
+
+        // for subfields, we only register one event so that we have access to the repeatable container name.
+        // all the uploaders for a given container are stored in the UploadersRepository.
+        if (! app('UploadersRepository')->hasUploadersRegisteredFor($uploader->getRepeatableContainerName())) {
+            $this->setupModelEvents($model, $uploader);
+        }
+
+        $subfields = collect($this->crudObject->getAttributes()['subfields']);
+        $subfields = $subfields->map(function ($item) use ($subfield, $uploader) {
+            if ($item['name'] === $subfield['name']) {
+                $item['upload'] = true;
+                $item['disk'] = $uploader->getDisk();
+                $item['prefix'] = $uploader->getPath();
+                if ($uploader->getTemporary()) {
+                    $item['temporary'] = $uploader->getTemporary();
+                    $item['expiration'] = $uploader->getExpiration();
+                }
+            }
+
+            return $item;
+        })->toArray();
+
+        app('UploadersRepository')->registerUploader($uploader->getRepeatableContainerName(), $uploader);
+
+        $this->crudObject->subfields($subfields);
+    }
+
+    /**
+     * Return the uploader for the object beeing configured.
+     * We will give priority to any uploader provided by `uploader => App\SomeUploaderClass` on upload definition.
+     *
+     * If none provided, we will use the Backpack defaults for the given object type.
+     *
+     * Throws an exception in case no uploader for the given object type is found.
+     *
+     * @param  array  $crudObject
+     * @param  array  $uploaderConfiguration
+     * @return UploaderInterface
+     *
+     * @throws Exception
+     */
+    private function getUploader(array $crudObject, array $uploaderConfiguration)
+    {
+        if (isset($uploaderConfiguration['uploaderType'])) {
+            return $uploaderConfiguration['uploaderType']::for($crudObject, $uploaderConfiguration);
+        }
+
+        if (app('UploadersRepository')->hasUploadFor($crudObject['type'], $this->macro)) {
+            return app('UploadersRepository')->getUploadFor($crudObject['type'], $this->macro)::for($crudObject, $uploaderConfiguration);
+        }
+
+        throw new Exception('Undefined upload type for '.$this->crudObjectType.' type: '.$crudObject['type']);
+    }
```

Comment:
I can't decide if this is any better of what we currently have:

```php
private function getUploader(array $crudObject, array $uploaderConfiguration)
    {
        $customUploader =  isset($uploaderConfiguration['uploaderType']) ?  $uploaderConfiguration['uploaderType']::for($crudObject, $uploaderConfiguration) : null;
        $stockUploader = app('UploadersRepository')->hasUploadFor($crudObject['type'], $this->macro) ? app('UploadersRepository')->getUploadFor($crudObject['type'], $this->macro)::for($crudObject, $uploaderConfiguration) : null;

        if (! $customUploader && ! $stockUploader) {
            throw new Exception('Undefined upload type for '.$this->crudObjectType.' type: '.$crudObject['type']);
        }
        return $customUploader ?? $stockUploader;
    }

// ALTERNATIVELY
private function getUploader(array $crudObject, array $uploaderConfiguration)
    {
        $customUploader =  isset($uploaderConfiguration['uploaderType']);
        $stockUploader = app('UploadersRepository')->hasUploadFor($crudObject['type'], $this->macro);

        if (! $customUploader && ! $stockUploader) {
            throw new Exception('Undefined upload type for '.$this->crudObjectType.' type: '.$crudObject['type']);
        }

       if($customUploader) {
            return $uploaderConfiguration['uploaderType']::for($crudObject, $uploaderConfiguration);
       }

       return app('UploadersRepository')->getUploadFor($crudObject['type'], $this->macro)::for($crudObject, $uploaderConfiguration)
    }
```

In our case the "default" is the exception because that function behaves like a "matcher", if no match found, exception is thrown. I think it's easier to read as we have now. 

You have the clear order:
- If you have a custom uploader, it's returned.
- If there is a stock uploader, it's returned.
- We got here? Exception, something is wrong.

Reply:
I think this one is easiest to read:
<img width="735" alt="CleanShot 2023-03-29 at 14 53 11@2x" src="https://user-images.githubusercontent.com/1032474/228541359-629210dc-0750-4d1a-b3f6-8029c296ddbd.png">

I don't like that the lines get so long. But other than that, it 100% is easier to read in my book. I don't even have to read the code itself, to my eyes it's just:

```php
private function getUploader(array $crudObject, array $uploaderConfiguration)
    {
        $customUploader =  'bla-bla';
        $stockUploader = 'some-other-bla';

        if (! $customUploader && ! $stockUploader) {
            throw new Exception('bla');
        }

        return $customUploader ?? $stockUploader;
    }
```

While the alternative requires a lot more brainpower to understand:

```php
private function getUploader(array $crudObject, array $uploaderConfiguration)
    {
        $customUploader =  'bla-bla';
        $stockUploader = 'some-other-bla';

        if (! $customUploader && ! $stockUploader) {
            throw new Exception('bla');
        }

       if($customUploader) {
            // ALERT - this does something important, I need to read it;
            return $uploaderConfiguration['uploaderType']::for($crudObject, $uploaderConfiguration);
       }

       // ALERT - this does something important, I need to read it;
       return app('UploadersRepository')->getUploadFor($crudObject['type'], $this->macro)::for($crudObject, $uploaderConfiguration)
    }
```

Reply:
Indeed, but in the example you most agreed on, we are creating "hidden rules". We transform booleans: **this exists? yes/no** into **this exists? this/null**.

Also you ALWAYS do two checks when the second check isn't relevant if the first one passes. This could be reflected on that example by chaining the coalescing operations, but that would become a mess to read imo. 

It's not expensive the second check in the example, but imagine that if instead of using a class in the container to hold the data (UploadersRepository) we were querying an api endpoint as our second check ($stockUploader in the example) ? 
```php
$custom = has custom check.
$stock = query the api.

return $custom ?? $stock. 
```

I would agree more with your approach if we were checking something to throw the exception. `if(smt) throw exception`. In that case I would 100% agree to kill the operation if the conditions are not met.
But initializing all the conditions to check for the exception first, I don't think it's the correct approach here.  

In our case I don't consider the exception the "default" behavior, that for me is described in the two if's. 
The exception is the `catch()` when nothing works, but we don't know beforehand what is needed to work, we just know that if we reach that point, we are out of the game.

Reply:
> Indeed, but in the example you most agreed on, we are creating "hidden rules". We transform booleans: this exists? yes/no into this exists? this/null.

Yes. You prioritize readability.

> Also you ALWAYS do two checks when the second check isn't relevant if the first one passes. This could be reflected on that example by chaining the coalescing operations, but that would become a mess to read imo. It's not expensive the second check in the example, but imagine that if instead of using a class in the container to hold the data (UploadersRepository) we were querying an api endpoint as our second check ($stockUploader in the example) ?

Fair enough. But then there are still ways to NOT end up with the alternative. If you want to return early, return early:

```php
private function getUploader(array $crudObject, array $uploaderConfiguration)
    {
        $customUploader =  'bla-bla';
        
        if ($customUploader) {
            return $customUploader;
        }

        $stockUploader = 'some-other-bla';

        if ($stockUploader) {
            return $stockUploader;
        }

        throw new Exception('bla');
    }
```

Shake hands on the above being both easy-to-read and performant?

Reply:
Hey Cristian, sorry, not yet. 

Not arguing about the readability, I think it's clear as you put it here, as it is clear as I've wrote it before. 

This is exactly the same as we have now `(if -> return, if -> return, exception)`, with the difference that you create two extra variables with information that we never going to use again ?

If you we needed something like this:

```php
$custom = bla;
if ($custom)
{
    Class::method($pass, $the, $custom);
    //or 
    $this->doSomethingWith($custom)
}
``` 

I would 100% agree with you and instantiate the variables up-front. 

In the case you proposed it's just wasted resources for 0 gain in readability. 
It's pretty much the same because the if conditions are small enough to don't polute the whole function, if they were big, maybe with bit-wise operators or something like, I would also agree with you into having a variable for it even if we are not going to re-use the variable again. 

In this case there is not a single benefit, not even readability improves in a way we can say it's worth it. Not really sure if it improves at all.

Let me know what you think. I know for sure that two variables does not have performance critical impact, but adding up stuff like this in a WHOLE app, will certainly starting to leave a memory footprint behind, even if later is garbage-collected.

Reply:
New rule: **If it isn't clear what a line or codeblock does by GLANCING IT (not reading it! just glancing it), then it needs a variable or a comment.** We should be able to read each function without having to read the actual logic. So yes, from now on, we consider readability more important than the performance gained by removing a one-off variable.